pf/tests/internal/testprovider/tuple_type.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testprovider
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+
+	"github.com/pulumi/pulumi-terraform-bridge/pf/internal/pfutils"
+)
+
+type TupleType struct {
```

Comment:
So if I understand this right this code is needed to even use tuples in PF via CustomType hooks.. So users are expected to write this up. Very interesting.

Reply:
I think tuples are less supported than possible right now. Defining our own tuple lets us test our ability to handle the tuple wire format and test support custom types at the same time.